src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs

```diff
@@ -15,30 +15,38 @@ internal partial class EndpointHtmlRenderer
 {
     private static readonly object ComponentSequenceKey = new object();
 
-    protected override IComponent ResolveComponentForRenderMode([DynamicallyAccessedMembers(Component)] Type componentType, int? parentComponentId, IComponentActivator componentActivator, IComponentRenderMode renderMode)
+    protected override IComponentRenderMode? ResolveEffectiveRenderMode(Type componentType, int? parentComponentId, IComponentRenderMode? componentTypeRenderMode, IComponentRenderMode? callerSpecifiedRenderMode)
     {
         if (_isHandlingErrors)
         {
             // Ignore the render mode boundary in error scenarios.
-            return componentActivator.CreateInstance(componentType);
+            return null;
         }
-        var closestRenderModeBoundary = parentComponentId.HasValue
-            ? GetClosestRenderModeBoundary(parentComponentId.Value)
-            : null;
 
-        if (closestRenderModeBoundary is not null)
+        // If we're inside a subtree with a rendermode, we don't need to emit another rendermode boundary.
+        // Once the subtree becomes interactive, the entire DOM subtree will get replaced anyway.
+        if (parentComponentId.HasValue && GetClosestRenderModeBoundary(parentComponentId.Value) is not null)
         {
-            // We're already inside a subtree with a rendermode. Once it becomes interactive, the entire DOM subtree
-            // will get replaced anyway. So there is no point emitting further rendermode boundaries.
-            return componentActivator.CreateInstance(componentType);
+            return null;
         }
-        else
+
+        if (_suppressRootComponentRenderModes && IsRootComponent(parentComponentId))
         {
-            // This component is the start of a subtree with a rendermode, so introduce a new rendermode boundary here
-            return new SSRRenderModeBoundary(_httpContext, componentType, renderMode);
+            // Disregard the callsite rendermode because this is in the root component (or is the root component itelf)
+            // and we've been configured to suppress render modes for the root component.
+            callerSpecifiedRenderMode = null;
```

Comment:
Isn't this logic disabling interactivity for all root components? Won't this mean that if there is an interactive component elsewhere in your app component it won't get interactivity?

Reply:
No, the `IsRootComponent` check avoids that.